Integrate Socket IO with Frontend

bharathcs opened the pull request:
# Add Jotai atoms and Socket.IO integration around Innkeeper

Elroy & Kevin wanted to use Jotai, so I looked up that stuff. Since a lot of the performance benefits come from
using Jotai utilities to neatly separate read / write dependencies down to the nesting / filtering, it would make
the most sense to abstract away the state management (including Socket IO work) away from the room page.

What works rn: Use the room page to connect, make matching request, and update the text editor. Run postman and act as the collaborator to make a matching request to complete the pair, and to watch the updates sent out by the browser. Remember to use something other than "user_a" in Postman.

---

## How this works.

### Atoms

Jotai manages state and automatically tracks read / write dependencies to avoid too many renders.

Basic Jotai atoms have been constructed, with the largest of which being roomState.
Dependent Jotai atoms have been constructed to filter / nest / focus on the correct values inside roomState.

### Socket IO

Handler is set up using a hook so that the Jotai-managed socket object will be initialised when the component loads,
and disconnected and set to null when the component unloads. (i.e. when /room page loads and is closed)

In the hook, handlers are attached to listen to server events and update Jotai-managed state (Jotai atoms of roomState, isConnected etc).
Additional Jotai atom was added as a utility function to support emitting events as simply as possible from the frontend

### Room Page

To work with how dumb the innkeeper auth logic is, a hardcoded 'user_a' is sent as auth token, and this is used to set up socket.

- If not connected, shows a loading page.
- If not matched, shows a placeholder lobby question (easy, medium, hard)
- Once matched, shows the question pane and code editor pane.

Use console and a Postman to verify that all the messages are being sent as expected.

---

- Add socket io deps
- Add socket io lib and share types
- Update CodeEditor for basic use
- Add jotai
- Fix build error
- Update innkeeper api type
- Fix bug from combining namespaces: use lobby room
- Add useInnkeeper hook and room-related Jotai atoms

bharathcs committed:
Add socket io deps

bharathcs committed:
Add socket io lib and share types

bharathcs committed:
Update CodeEditor for basic use

bharathcs committed:
Add useInnkeeper hook and room-related Jotai atoms

bharathcs committed:
Add jotai

bharathcs committed:
Update innkeeper api type

bharathcs committed:
Fix build error

bharathcs commented on frontend/package.json:
Had to add these additional npm pkgs: `bufferutil, optics-ts, utf-8-validate, ws` for some reason. I kept having build errors until all were added.

sivayogasubramanian replied:
You need to add the new env variable to the CI files as well as the github secrets

dependabot[bot] committed:
chore(deps): bump eslint-config-next from 13.4.19 to 13.5.3 in /frontend

Bumps [eslint-config-next](https://github.com/vercel/next.js/tree/HEAD/packages/eslint-config-next) from 13.4.19 to 13.5.3.
- [Release notes](https://github.com/vercel/next.js/releases)
- [Changelog](https://github.com/vercel/next.js/blob/canary/release.js)
- [Commits](https://github.com/vercel/next.js/commits/v13.5.3/packages/eslint-config-next)

---
updated-dependencies:
- dependency-name: eslint-config-next
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dependabot[bot] committed:
chore(deps): bump @types/react from 18.2.21 to 18.2.23 in /frontend

Bumps [@types/react](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/HEAD/types/react) from 18.2.21 to 18.2.23.
- [Release notes](https://github.com/DefinitelyTyped/DefinitelyTyped/releases)
- [Commits](https://github.com/DefinitelyTyped/DefinitelyTyped/commits/HEAD/types/react)

---
updated-dependencies:
- dependency-name: "@types/react"
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

sivayogasubramanian commented on frontend/.env.local.example:
```diff
@@ -6,4 +6,5 @@ NEXT_PUBLIC_FIREBASE_PROJECT_ID=
-NEXT_PUBLIC_GOOGLE_CLIENT_ID=
\ No newline at end of file
+NEXT_PUBLIC_GOOGLE_CLIENT_ID=
+NEXT_PUBLIC_PEERPREP_INNKEEPER_SOCKET_URL=
```
You need to add the new env variable to the CI files as well as the github secrets

bharathcs replied:
Thanks for the catch: [Update env for github workflows](https://github.com/PeerPrep/peerprep/pull/53/commits/39da179f508d34ad6befceada5f9d8f323076885)

dependabot[bot] committed:
chore(deps): bump daisyui from 3.8.0 to 3.8.2 in /frontend

Bumps [daisyui](https://github.com/saadeghi/daisyui) from 3.8.0 to 3.8.2.
- [Changelog](https://github.com/saadeghi/daisyui/blob/master/CHANGELOG.md)
- [Commits](https://github.com/saadeghi/daisyui/compare/v3.8.0...v3.8.2)

---
updated-dependencies:
- dependency-name: daisyui
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

bharathcs committed:
Update env for github workflows
